File: iso27001-fastapi/app/infrastructure/quality_score.py
"""
Risk-weighted Quality Score Calculator.

Composite score derived from five pillars aligned to ISO 27001 domains:

  Pillar           Weight   ISO 27001 Annex
  ─────────────────────────────────────────
  Security          40%     A.9, A.10
  Data Integrity    20%     A.12
  Reliability       15%     A.17
  Auditability      15%     A.12
  Performance        5%     A.17
  (Bonus gap)        5%     reserved

Score of 1.0 = perfect compliance; 0.0 = complete failure.
A score below 0.70 should block production deployments.
"""
from __future__ import annotations

from dataclasses import dataclass


PRODUCTION_GATE = 0.70   # minimum acceptable score


@dataclass(frozen=True)
class SloAlert:
    """
    A.17: SLO breach signals for alerting pipelines.

    Fields are set to True when the corresponding SLO threshold is exceeded.
    All four signals are independent — any True value should trigger an alert.
    """
    p95_latency_breached: bool   # observed P95 > SLO_P95_LATENCY_MS
    p99_latency_breached: bool   # observed P99 > SLO_P99_LATENCY_MS
    error_rate_breached: bool    # 5xx error rate > SLO_ERROR_RATE_PCT
    client_error_spike: bool     # 4xx rate > CLIENT_ERROR_SPIKE_PCT (abuse signal)

    def any_breach(self) -> bool:
        """True if at least one SLO is currently violated."""
        return (
            self.p95_latency_breached
            or self.p99_latency_breached
            or self.error_rate_breached
            or self.client_error_spike
        )

    def to_dict(self) -> dict[str, bool]:
        return {
            "any_breach": self.any_breach(),
            "p95_latency_breached": self.p95_latency_breached,
            "p99_latency_breached": self.p99_latency_breached,
            "error_rate_breached": self.error_rate_breached,
            "client_error_spike": self.client_error_spike,
        }


@dataclass(frozen=True)
class QualityScore:
    security: float          # 0–1  (JWT validity, RBAC enforcement, encryption)
    data_integrity: float    # 0–1  (audit log completeness, event dispatch success)
    reliability: float       # 0–1  (1 - error_rate, derived from ErrorBudgetTracker)
    auditability: float      # 0–1  (structured log coverage, correlation ID presence)
    performance: float       # 0–1  (P95 latency vs SLA threshold)

    def composite(self) -> float:
        """
        Weighted composite score, normalized to [0.0, 1.0].
        Weights sum to 0.95 (5% reserved for a future gap pillar);
        dividing by 0.95 maps a perfect score to exactly 1.0.
        """
        _WEIGHT_SUM = 0.95
        raw = (
            self.security        * 0.40
            + self.data_integrity  * 0.20
            + self.reliability     * 0.15
            + self.auditability    * 0.15
            + self.performance     * 0.05
        )
        return raw / _WEIGHT_SUM

    def passes_gate(self) -> bool:
        """True if this score meets the production deployment gate."""
        return self.composite() >= PRODUCTION_GATE

    def to_dict(self) -> dict[str, object]:
        c = self.composite()
        return {
            "composite": round(c, 4),
            "passes_gate": self.passes_gate(),
            "production_gate_threshold": PRODUCTION_GATE,
            "pillars": {
                "security":       {"score": self.security,       "weight": 0.40},
                "data_integrity": {"score": self.data_integrity, "weight": 0.20},
                "reliability":    {"score": self.reliability,    "weight": 0.15},
                "auditability":   {"score": self.auditability,   "weight": 0.15},
                "performance":    {"score": self.performance,    "weight": 0.05},
            },
        }


class QualityScoreCalculator:
    """
    Assembles a QualityScore and SloAlert from live runtime signals.

    Inject the dependencies it needs (error budget, latency histogram, etc.)
    rather than reaching for global state directly — makes it unit-testable.
    """

    # A.17: SLO thresholds — defined once, referenced everywhere.
    SLO_P95_LATENCY_MS: float = 200.0    # alert if P95 exceeds 200 ms
    SLO_P99_LATENCY_MS: float = 500.0    # alert if P99 exceeds 500 ms
    SLO_ERROR_RATE_PCT: float = 0.1      # alert if 5xx rate (%) exceeds 0.1% (= 99.9% SLA)
    CLIENT_ERROR_SPIKE_PCT: float = 5.0  # alert if 4xx rate (%) exceeds 5% (abuse signal)

    def __init__(
        self,
        *,
        sla_latency_p95_ms: float = 200.0,
        target_latency_ms: float = 500.0,
        sla_latency_p99_ms: float = 0.0,
    ) -> None:
        """
        Args:
            sla_latency_p95_ms: Observed P95 latency in milliseconds (from metrics).
            target_latency_ms:  Acceptable P95 latency ceiling from SLA.
            sla_latency_p99_ms: Observed P99 latency in milliseconds (from metrics).
        """
        self._sla_latency_p95_ms = sla_latency_p95_ms
        self._target_latency_ms = target_latency_ms
        self._sla_latency_p99_ms = sla_latency_p99_ms

    def calculate(
        self,
        *,
        auth_checks_passed: int,
        auth_checks_total: int,
        audit_events_recorded: int,
        audit_events_expected: int,
        availability: float,
        logs_with_correlation_id: int,
        total_logs: int,
    ) -> QualityScore:
        """
        Compute each pillar score and return a QualityScore.

        All inputs must be non-negative integers or floats in [0, 1].
        """
        security = self._ratio(auth_checks_passed, auth_checks_total)
        data_integrity = self._ratio(audit_events_recorded, audit_events_expected)
        reliability = max(0.0, min(1.0, availability))
        auditability = self._ratio(logs_with_correlation_id, total_logs)
        performance = self._latency_score()

        return QualityScore(
            security=security,
            data_integrity=data_integrity,
            reliability=reliability,
            auditability=auditability,
            performance=performance,
        )

    def slo_alert(
        self,
        *,
        failed_requests: int = 0,
        client_errors: int = 0,
        total_requests: int = 0,
    ) -> SloAlert:
        """
        A.17: Evaluate SLO breach conditions.

        Returns an SloAlert whose fields indicate which thresholds are violated.
        Wire this into your alerting pipeline: any_breach() == True → fire alert.
        """
        error_rate_pct = (failed_requests / total_requests * 100.0) if total_requests > 0 else 0.0
        client_error_pct = (client_errors / total_requests * 100.0) if total_requests > 0 else 0.0

        return SloAlert(
            p95_latency_breached=self._sla_latency_p95_ms > self.SLO_P95_LATENCY_MS,
            p99_latency_breached=self._sla_latency_p99_ms > self.SLO_P99_LATENCY_MS,
            error_rate_breached=error_rate_pct > self.SLO_ERROR_RATE_PCT,
            client_error_spike=client_error_pct > self.CLIENT_ERROR_SPIKE_PCT,
        )

    # ── helpers ──────────────────────────────────────────────────────────────

    @staticmethod
    def _ratio(numerator: int, denominator: int) -> float:
        if denominator <= 0:
            return 1.0   # no data = assume perfect (avoid false negatives at startup)
        return max(0.0, min(1.0, numerator / denominator))

    def _latency_score(self) -> float:
        """
        Linear decay: 1.0 at 0ms, 0.0 at target_latency_ms and beyond.
        """
        if self._target_latency_ms <= 0:
            return 1.0
        ratio = self._sla_latency_p95_ms / self._target_latency_ms
        return max(0.0, 1.0 - ratio)

File: iso27001-phoenix/lib/iso27001_phoenix/infrastructure/telemetry/quality_score.ex
defmodule Iso27001Phoenix.Infrastructure.Telemetry.QualityScore do
  @moduledoc """
  A.17: Risk-weighted quality score calculator.

  Pillar           Weight   ISO 27001 Annex
  ─────────────────────────────────────────
  Security          40%     A.9, A.10
  Data Integrity    20%     A.12
  Reliability       15%     A.17
  Auditability      15%     A.12
  Performance        5%     A.17
  (Gap bonus)        5%     reserved

  Score 1.0 = perfect; gate threshold = 0.70.
  """

  @production_gate 0.70
  @weight_sum 0.95

  # SLO thresholds — A.17
  @slo_p95_latency_ms 200.0
  @slo_p99_latency_ms 500.0
  @slo_error_rate_pct 0.1
  @client_error_spike_pct 5.0

  defstruct [:security, :data_integrity, :reliability, :auditability, :performance]

  @doc "Computes the weighted composite score."
  def composite(%__MODULE__{} = s) do
    raw =
      s.security       * 0.40 +
      s.data_integrity * 0.20 +
      s.reliability    * 0.15 +
      s.auditability   * 0.15 +
      s.performance    * 0.05

    raw / @weight_sum
  end

  def passes_gate?(%__MODULE__{} = s), do: composite(s) >= @production_gate

  @doc "Builds a QualityScore from runtime signals."
  def calculate(opts) do
    auth_passed   = Keyword.get(opts, :auth_checks_passed, 1)
    auth_total    = Keyword.get(opts, :auth_checks_total, 1)
    audit_rec     = Keyword.get(opts, :audit_events_recorded, 1)
    audit_exp     = Keyword.get(opts, :audit_events_expected, 1)
    availability  = Keyword.get(opts, :availability, 1.0)
    logs_with_cid = Keyword.get(opts, :logs_with_correlation_id, 1)
    total_logs    = Keyword.get(opts, :total_logs, 1)
    p95_ms        = Keyword.get(opts, :p95_latency_ms, 0.0)

    %__MODULE__{
      security:       ratio(auth_passed, auth_total),
      data_integrity: ratio(audit_rec, audit_exp),
      reliability:    max(0.0, min(1.0, availability)),
      auditability:   ratio(logs_with_cid, total_logs),
      performance:    latency_score(p95_ms)
    }
  end

  @doc "Returns SLO breach flags."
  def slo_alert(opts \\ []) do
    failed  = Keyword.get(opts, :failed_requests, 0)
    clients = Keyword.get(opts, :client_errors, 0)
    total   = Keyword.get(opts, :total_requests, 0)
    p95_ms  = Keyword.get(opts, :p95_latency_ms, 0.0)
    p99_ms  = Keyword.get(opts, :p99_latency_ms, 0.0)

    error_rate_pct  = if total > 0, do: failed / total * 100.0, else: 0.0
    client_pct      = if total > 0, do: clients / total * 100.0, else: 0.0

    %{
      p95_latency_breached: p95_ms > @slo_p95_latency_ms,
      p99_latency_breached: p99_ms > @slo_p99_latency_ms,
      error_rate_breached:  error_rate_pct > @slo_error_rate_pct,
      client_error_spike:   client_pct > @client_error_spike_pct,
      any_breach: p95_ms > @slo_p95_latency_ms or p99_ms > @slo_p99_latency_ms or
                  error_rate_pct > @slo_error_rate_pct or client_pct > @client_error_spike_pct
    }
  end

  defp ratio(_, 0), do: 1.0
  defp ratio(n, d), do: max(0.0, min(1.0, n / d))

  defp latency_score(p95_ms) do
    target = @slo_p95_latency_ms
    max(0.0, 1.0 - p95_ms / target)
  end
end
